src/main/scala/vectorpipe/osm/internal/functions/package.scala

```diff
@@ -0,0 +1,676 @@
+        env.expandToInclude(getX(i), getY(i))
+      }
+
+      env
+    }
+
+    override def clone(): AnyRef = new PartialCoordinateSequence(sequence, offset)
+  }
+
+  // rather than being a nested set of CoordinateSequences, this is a mutable wrapper to avoid deep call stacks
+  class VirtualCoordinateSequence(sequences: Seq[CoordinateSequence]) extends CoordinateSequence {
+    private val rangeMap: RangeMap[Integer, CoordinateSequence] = TreeRangeMap.create[Integer, CoordinateSequence]
+
+    sequences.zip(sequences.map(_.size).scanLeft(0)(_ + _).dropRight(1))
+      .map { case (seq, offset) => (seq, Range.closed(offset: Integer, offset + seq.size - 1: Integer)) }
+      .foreach { case (seq, range) => rangeMap.put(range, seq)}
+
+    private var dimension: Int = sequences.map(_.getDimension).min
+
+    private var _size: Int = sequences.map(_.size).sum
+
+    private def getSequence(i: Int): (CoordinateSequence, Int) = {
+      val entry = rangeMap.getEntry(i: Integer)
+
+      (entry.getValue, i - entry.getKey.lowerEndpoint)
+    }
+
+    def append(sequence: CoordinateSequence): VirtualCoordinateSequence = {
+      val upperEndpoint = rangeMap.span.upperEndpoint
+      val range = Range.closed(upperEndpoint + 1: Integer, upperEndpoint + sequence.size: Integer)
+      rangeMap.put(range, sequence)
+
+      dimension = Math.min(dimension, sequence.getDimension)
+      _size += sequence.size
+
+      this
+    }
+
+    override def getDimension: Int = dimension
+
+    override def getCoordinate(i: Int): Coordinate = {
+      val (sequence, index) = getSequence(i)
+
+      // bypass PackedCoordinateSequence.getCoordinate to prevent caching and associated allocation
+      new Coordinate(sequence.getX(index), sequence.getY(index))
+    }
+
+    override def getCoordinateCopy(i: Int): Coordinate = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getCoordinateCopy(index)
+    }
+
+    override def getCoordinate(i: Int, coord: Coordinate): Unit = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getCoordinate(index, coord)
+    }
+
+    override def getX(i: Int): Double = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getX(index)
+    }
+
+    override def getY(i: Int): Double = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getY(index)
+    }
+
+    override def getOrdinate(i: Int, ordinateIndex: Int): Double = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getOrdinate(index, ordinateIndex)
+    }
+
+    override def size(): Int = _size
+
+    override def setOrdinate(i: Int, ordinateIndex: Int, value: Double): Unit = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.setOrdinate(index, ordinateIndex, value)
+    }
+
+    // TODO this should be invalidated after append (but it doesn't actually matter because all of the appending will
+    // occur ahead of time)
+    private lazy val coordinates: Array[Coordinate] = {
```

Comment:
This is one of those fast and loose spots where invalidating is skipped because we know `coordinates` will only be requested after the sequence has been fully assembled.